### ISSUE TITLE
Page Setting Custom Controls

### ISSUE THREAD

**Issue by ryanlabelle** [pojome/elementor#3928]
## Description
Thank you for reviewing my question. 

We use to be able to add controls to Page Settings using the code below. As for Elementor 2.0 custom controls no longer show up.

## Steps to reproduce

```
if ( ! function_exists( 'th_add_custom_controls_elem_page_settings_bottom' ) ) {
    function th_add_custom_controls_elem_page_settings_bottom( \Elementor\Core\Settings\Page\Model $page )
    {

        if(isset($page) && $page->get_id() > "") {
            $th_post_type = false;
            $th_post_type = get_post_type($page->get_id());
            if ($th_post_type == 'page' || $th_post_type == 'themo_tour' || $th_post_type == 'themo_portfolio' || $th_post_type == 'themo_room') {

                $page->add_control(
                    'themo_page_layout',
                    [
                        'label' => __( 'Sidebar', 'th-widget-pack' ),
                        'type' => \Elementor\Controls_Manager::CHOOSE,
                        'default' => 'full',
                        'options' => [
                            'left'    => [
                                'title' => __( 'Left', 'th-widget-pack' ),
                                'icon' => 'fa fa-long-arrow-left',
                            ],
                            'full' => [
                                'title' => __( 'No Sidebar', 'th-widget-pack' ),
                                'icon' => 'fa fa-times',
                            ],
                            'right' => [
                                'title' => __( 'Right', 'th-widget-pack' ),
                                'icon' => 'fa fa-long-arrow-right',
                            ],

                        ],
                        'return_value' => 'yes',
                    ]
                );
            }
        }

    }
}
add_action( 'elementor/element/page-settings/section_page_settings/after_section_start', 'th_add_custom_controls_elem_page_settings_top',10, 2);
```

Looks like Document Settings section https://cl.ly/1p3S2Y0L0b2F has been changed to General Settings https://cl.ly/2g3u0t0b3o2G

## Environment

Elemenotor 2.0

**Comment by bainternet** [comment]
@ryanlabelle 

The hook name was changed in v2.0.0 when we started using document types, try:
`elementor/element/post/document_settings/after_section_start`

**Comment by ryanlabelle** [comment]
Thanks @bainternet  👍

**Comment by notrealdev** [comment]
Not worked for me

**Comment by ryanlabelle** [comment]
@notrealdev this worked for me:

```
// Add custom controls to the Page Settings inside the Elementor Global Options.

// Top of section
if ( ! function_exists( 'th_add_custom_controls_elem_page_settings_top' ) ) {
    function th_add_custom_controls_elem_page_settings_top(\Elementor\Core\DocumentTypes\Post $page)
    {

        if(isset($page) && $page->get_id() > ""){
            $th_post_type = false;
            $th_post_type = get_post_type($page->get_id());
            if($th_post_type == 'page' || $th_post_type == 'themo_tour' || $th_post_type == 'themo_portfolio' || $th_post_type == 'themo_room' || $th_post_type == 'themo_hole'){

                $page->add_control(
                    'themo_transparent_header',
                    [
                        'label' => __( 'Transparent Header', 'th-widget-pack' ),
                        'type' => \Elementor\Controls_Manager::SWITCHER,
                        'default' => 'Off',
                        'label_on' => __( 'On', 'th-widget-pack' ),
                        'label_off' => __( 'Off', 'th-widget-pack' ),
                        'return_value' => 'on',
                    ]
                );

                $page->add_control(
                    'themo_header_content_style',
                    [
                        'label' => __( 'Transparent Header Content Style', 'th-widget-pack' ),
                        'type' => \Elementor\Controls_Manager::SELECT,
                        'label_block' => true,
                        'default' => 'light',
                        'options' => [
                            'light' => __( 'Light', 'th-widget-pack' ),
                            'dark' => __( 'Dark', 'th-widget-pack' ),
                        ],
                        'condition' => [
                            'themo_transparent_header' => 'on',
                        ],
                    ]
                );

                $page->add_control(
                    'themo_alt_logo',
                    [
                        'label' => __( 'Use Alternative Logo', 'th-widget-pack' ),
                        'description' => __( 'You can upload an alternative logo under Appearance / Customize / Theme Options / Logo / ', 'th-widget-pack' ),
                        'type' => \Elementor\Controls_Manager::SWITCHER,
                        'default' => 'Off',
                        'label_on' => __( 'On', 'th-widget-pack' ),
                        'label_off' => __( 'Off', 'th-widget-pack' ),
                        'return_value' => 'on',
                        'condition' => [
                            'themo_transparent_header' => 'on',
                        ],
                    ]
                );

                $page_title_selector = get_option( 'elementor_page_title_selector' );
                if ( empty( $page_title_selector ) ) {
                    $page_title_selector = 'h1.entry-title';
                }


                $page->add_control(
                    'themo_page_title_margin',
                    [
                        'label' => __( 'Title  Margin', 'th-widget-pack' ),
                        'type' => \Elementor\Controls_Manager::SLIDER,
                        'default' => [
                            'size' => 1,
                        ],
                        'range' => [
                            'px' => [
                                'min' => 0,
                                'max' => 1000,
                                'step' => 5,
                            ],
                            '%' => [
                                'min' => 0,
                                'max' => 100,
                            ],
                        ],
                        'size_units' => [ 'px', '%' ],
                        'selectors' => [
                            '{{WRAPPER}} ' . $page_title_selector => 'margin-top: {{SIZE}}{{UNIT}};',
                        ],
                    ]
                );
            }
        }

    }
}
// Bottom of section
if ( ! function_exists( 'th_add_custom_controls_elem_page_settings_bottom' ) ) {
    function th_add_custom_controls_elem_page_settings_bottom( \Elementor\Core\DocumentTypes\Post $page )
    {

        if(isset($page) && $page->get_id() > "") {
            $th_post_type = false;
            $th_post_type = get_post_type($page->get_id());
            if ($th_post_type == 'page' || $th_post_type == 'themo_tour' || $th_post_type == 'themo_portfolio' || $th_post_type == 'themo_room' || $th_post_type == 'themo_hole') {

                $page->add_control(
                    'themo_page_layout',
                    [
                        'label' => __( 'Sidebar', 'th-widget-pack' ),
                        'type' => \Elementor\Controls_Manager::CHOOSE,
                        'default' => 'full',
                        'options' => [
                            'left'    => [
                                'title' => __( 'Left', 'th-widget-pack' ),
                                'icon' => 'fa fa-long-arrow-left',
                            ],
                            'full' => [
                                'title' => __( 'No Sidebar', 'th-widget-pack' ),
                                'icon' => 'fa fa-times',
                            ],
                            'right' => [
                                'title' => __( 'Right', 'th-widget-pack' ),
                                'icon' => 'fa fa-long-arrow-right',
                            ],

                        ],
                        'return_value' => 'yes',
                    ]
                );
            }
        }

    }
}
add_action( 'elementor/element/post/document_settings/after_section_start', 'th_add_custom_controls_elem_page_settings_top',10, 2);
add_action( 'elementor/element/post/document_settings/before_section_end', 'th_add_custom_controls_elem_page_settings_bottom',10, 2);
```
Then I added some JS to update and reload the Elementor Editor Window.

```
/**
 * Created by rl on 2017-03-30.
 */
jQuery( function( $ ) {

    // Page Settings Panel - onchange save and reload elementor window.

    if ( typeof elementor != "undefined" && typeof elementor.settings.page != "undefined") {

        // Page Layout Options
        elementor.settings.page.addChangeCallback( 'themo_page_layout', function( newValue ) {
            // Here you can do as you wish with the newValue

            //console.log('PAGE LAYOUT CHANGE.');

            elementor.saver.update( {
                onSuccess: function() {

                    //console.log('SAVE');
                    elementor.reloadPreview();

                    elementor.once( 'preview:loaded', function() {
                        elementor.getPanelView().setPage( 'page_settings' );
                    } );
                }
            } );
        } );

        // Header Transparency
        elementor.settings.page.addChangeCallback( 'themo_transparent_header', function( newValue ) {
            // Here you can do as you wish with the newValue

            elementor.saver.update( {
                onSuccess: function() {

                    //console.log('SAVE');
                    elementor.reloadPreview();

                    elementor.once( 'preview:loaded', function() {
                        elementor.getPanelView().setPage( 'page_settings' );
                    } );
                }
            } );



        } );

        // Header Contenet Style
        elementor.settings.page.addChangeCallback( 'themo_header_content_style', function( newValue ) {
            // Here you can do as you wish with the newValue

            elementor.saver.update( {
                onSuccess: function() {

                    //console.log('SAVE');
                    elementor.reloadPreview();

                    elementor.once( 'preview:loaded', function() {
                        elementor.getPanelView().setPage( 'page_settings' );
                    } );
                }
            } );
        } );

        // Alt Logo
        elementor.settings.page.addChangeCallback( 'themo_alt_logo', function( newValue ) {
            // Here you can do as you wish with the newValue

            //console.log('NEW VALUE ALT LOGO '+newValue);

            elementor.saver.update( {
                onSuccess: function() {

                    //console.log('SAVE');
                    elementor.reloadPreview();

                    elementor.once( 'preview:loaded', function() {
                        elementor.getPanelView().setPage( 'page_settings' );
                    } );
                }
            } );
        } );

    }


    if ( undefined !== window.elementor ) {

        elementor.hooks.addAction('panel/open_editor/widget', function (panel, model, view) {

            //elementor.reloadPreview();

            /*var $element = view.$el.find( '.elementor-selector' );

             if ( $element.length ) {
             $element.click( function() {
             alert( 'Some Message' );
             } );
             }*/

            /* var self = this;

             var settings = self.model.toJSON();

             settings.id = elementor.config.post_id;

             NProgress.start();

             elementor.ajax.send( 'save_page_settings', {
             data: settings,
             success: function() {
             NProgress.done();

             self.setSettings( 'savedSettings', settings );

             self.hasChange = false;

             if ( callback ) {
             callback.apply( self, arguments );
             }
             console.log('saved');
             },
             error: function() {
             alert( 'An error occurred' );
             }
             } );*/

        });
    }
} );
```

**Comment by notrealdev** [comment]
Thank @ryanlabelle it worked, and one more question, how can I create a new section like this? via hook? is it feasible?

![section](https://i.imgur.com/57SFGIS.png)

**Comment by ryanlabelle** [comment]
@notrealdev sorry I don't have code for that yet. :)

**Comment by avitorio** [comment]
@notrealdev I know this was posted a long time ago, but I've managed to achieve what you want with this code:

`add_action( 'elementor/element/post/document_settings/after_section_end', 'page_sections_controls',10, 2);

// Add new controls section
if ( ! function_exists( 'page_sections_controls' ) ) {
    function page_sections_controls( \Elementor\Core\DocumentTypes\Post $page )
    {

        if(isset($page) && $page->get_id() > "") {
            $th_post_type = false;
			$th_post_type = get_post_type($page->get_id());
            if ($th_post_type == 'page' || $th_post_type == 'revision') {
				
				// Header Options
				$page->start_controls_section(
					'header_section',
					[
						'label' => __( 'Header', 'molotov-form' ),
						'tab' => \Elementor\Controls_Manager::TAB_SETTINGS,
					]
				);
		
				$form_description = '<p>Hello></p>';
		
				$page->add_control(
					'anchor_description',
					[
						'raw' => __( $form_description, 'molotov-form' ),
						'type' => \Elementor\Controls_Manager::RAW_HTML,
					]
				);
		
				$page->end_controls_section();
            }
        }

    }
}`

**Comment by riadcitaku** [comment]
@avitorio is this still working? I have tested it and got the following error:

`Elementor: You can't start a section before the end of the previous section "document_settings".`

**Comment by avitorio** [comment]
@riadcitaku not sure, the install where this was working still runs an old version of elementor.

Are you using the 'after_section_end' hook? From the error it seems you are starting a section halfway through another one.

**Comment by riadcitaku** [comment]
@avitorio it seems like I was using 'before_section_end', thank you for the reply 👍

**Comment by Paras16699** [comment]
@bainternet You guys might want to update this in the developer docs here which still has that old code example - https://developers.elementor.com/elementor-document-settings/

**Comment by Paras16699** [comment]
Just realized that `post` has now been changed to `wp-post` since v2.7.0

Please do regularly update the developers' docs when making changes to hooks and actions.

**Comment by oooh-boi** [comment]
OK... for all of you stuck with how to add control(s) to document Settings here's what worked for me after 6 hours of trial and error party found under this thread...

The hook:
`add_action( 'elementor/element/post/document_settings/before_section_end', [ $this, 'YOUR_FUNCTION' ], 10, 2 );`

The function:
```
public function YOUR_FUNCTION( \Elementor\Core\DocumentTypes\PageBase $page ) {
	$page->add_control(
		'control_id',
		[
			/* ... controls ... */
		]
	);
}
```
**The above is how controls work for both Page and Post post types.**

If you just pass over:
`\Elementor\Core\DocumentTypes\Page`
...or...
`\Elementor\Core\DocumentTypes\Post`
...it'll end up with an error for either of these two counterparts been currently edit in Elementor.

**Comment by rhino-corey** [comment]
None of the methods above work for me anymore. I had to use this hook and method signature:

`add_action( 'elementor/documents/register_controls', [$this, 'register_page_controls'] );`

```
public function register_page_controls( $document ){
	if ( ! $document instanceof \Elementor\Core\DocumentTypes\PageBase || ! $document::get_property( 'has_elements' ) ) {
		return;
	}
		
	$document->start_controls_section(
		'page_section',
		[
			'label' => esc_html__( 'New Section', 'my-domain' ),
			'tab' => \Elementor\Controls_Manager::TAB_SETTINGS,
		]
	);
	
	$document->add_control(
		'page_feature_enabled',
		[
			'label' => __( 'Enabled', 'my-domain' ),
			'type' => \Elementor\Controls_Manager::SWITCHER,
			'default' => 'no',
			'label_on' => __( 'Yes', 'my-domain' ),
			'label_off' => __( 'No', 'my-domain' ),
			'return_value' => 'yes',
		]
	);
	
	$document->end_controls_section();
}
```

Ref: https://developers.elementor.com/docs/editor/page-settings-panel/
